agent/engine/dockerstate/dockerstate_test.go

```diff
@@ -361,3 +363,93 @@ func TestTaskIPAddress(t *testing.T) {
 	assert.True(t, ok)
 	assert.Equal(t, addr, taskIP)
 }
+
+// TestGenerateAndAddV3EndpointIDAddNewIDSucceeds tests that when the container didn't
+// have a v3 endpoint id and we successfully generated a new uuid, the function returns success and
+// the id is saved and added to relevant maps
+func TestGenerateAndAddV3EndpointIDAddNewIDSucceeds(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	state := newDockerTaskEngineState()
+	assert.NotNil(t, state.v3EndpointIDToTask)
+	assert.NotNil(t, state.v3EndpointIDToContainerName)
+
+	container := &apicontainer.Container{
+		Name: "containerName",
+	}
+	task := &apitask.Task{
+		Arn: "taskArn",
+		Containers: []*apicontainer.Container{
+			container,
+		},
+	}
+	mockUUID := mock_uuidwrapper.NewMockUUID(ctrl)
+
+	mockUUID.EXPECT().New().Return("new-uuid")
+
+	assert.True(t, state.GenerateAndAddV3EndpointID(task, container, mockUUID))
+	assert.Equal(t, container.V3EndpointID, "new-uuid")
+	assert.Equal(t, state.v3EndpointIDToTask["new-uuid"], "taskArn")
+	assert.Equal(t, state.v3EndpointIDToContainerName["new-uuid"], "containerName")
+}
+
+// TestGenerateAndAddV3EndpointIDAlreadyExists tests that when the container already has
+// a v3 endpoint id, the function return success without generating a new id for it
+func TestGenerateAndAddV3EndpointIDAlreadyExists(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	state := newDockerTaskEngineState()
+	assert.NotNil(t, state.v3EndpointIDToTask)
+	assert.NotNil(t, state.v3EndpointIDToContainerName)
+
+	container := &apicontainer.Container{
+		Name:         "containerName",
+		V3EndpointID: "containerV3EndpointID",
+	}
+	task := &apitask.Task{
+		Arn: "taskArn",
+		Containers: []*apicontainer.Container{
+			container,
+		},
+	}
+	mockUUID := mock_uuidwrapper.NewMockUUID(ctrl)
+
+	assert.True(t, state.GenerateAndAddV3EndpointID(task, container, mockUUID))
+	assert.Equal(t, container.V3EndpointID, "containerV3EndpointID")
+}
+
+// TestGenerateAndAddV3EndpointIDFailure tests that when the container didn't
+// have a v3 endpoint id and we failed to generate a new uuid after maximum retries,
+// the function returns failure without setting the container's id
+func TestGenerateAndAddV3EndpointIDFailure(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	state := newDockerTaskEngineState()
+	assert.NotNil(t, state.v3EndpointIDToTask)
+	assert.NotNil(t, state.v3EndpointIDToContainerName)
+
+	container := &apicontainer.Container{
+		Name:         "containerName",
+		V3EndpointID: "",
+	}
+	task := &apitask.Task{
+		Arn: "taskArn",
+		Containers: []*apicontainer.Container{
+			container,
+		},
+	}
+	mockUUID := mock_uuidwrapper.NewMockUUID(ctrl)
+
+	state.v3EndpointIDToTask["containerV3EndpointID"] = "taskArn"
+	state.v3EndpointIDToContainerName["containerV3EndpointID"] = "containerName"
+
+	for i := 0; i < maximumGenerateIDRetries; i++ {
+		mockUUID.EXPECT().New().Return("containerV3EndpointID")
```

Comment:
A mock isn't necessary for this object. Just do something like this:

```golang
type StaticUUIDProvider struct {
    TestId string
}

func (t *StaticUUIDProvider) New() string {
    return t.TestId
}
```

It may also be possible to just use the uuid library directly in the test.